python/rmm/rmm.py

```diff
@@ -231,3 +239,40 @@ def rmm_cupy_allocator(nbytes):
     ptr = cupy.cuda.memory.MemoryPointer(mem, 0)
 
     return ptr
+
+
+def register_reinitialize_hook(func, *args, **kwargs):
+    """
+    Add a function to the list of functions that will be called before
+    `rmm.reinitialize()`.
+
+    Typically, a library will use this function to register hooks that
+    are responsible for deleting any remaining internal references to
+    objects using device memory allocated by RMM.
+
+    Hooks are called in the *reverse* order they are registered.
```

Comment:
While I don't think it is problematic, is this a guarantee that we want to offer to libraries registering hooks? I think that if a library _relies_ on this behaviour then it is likely too delicate. Not least because you (as a library) cannot know in which order your hook was added.

Thus, I would prefer if no such guarantee were offered.

Reply:
I could be persuaded that the order is important if reinitialisation is required to be collective across more than one process in some sense (but then we have to think harder about what the semantics need to be).